File: app/src/test/java/com/futsch1/medtimer/JSONBackupFuzzTest.java
package com.futsch1.medtimer;

import static org.mockito.Mockito.mock;

import android.util.Log;

import com.code_intelligence.jazzer.junit.FuzzTest;
import com.futsch1.medtimer.database.JSONBackup;
import com.futsch1.medtimer.database.JSONMedicineBackup;
import com.futsch1.medtimer.database.JSONReminderEventBackup;
import com.futsch1.medtimer.database.MedicineRepository;

import java.util.List;

@SuppressWarnings("java:S2187")
public class JSONBackupFuzzTest {
    @FuzzTest
    void fuzzTestMedicineBackup(String json) {
        JSONMedicineBackup jsonMedicineBackup = new JSONMedicineBackup();
        JSONReminderEventBackup jsonReminderEventBackup = new JSONReminderEventBackup();

        Log.enable = false;

        checkBackup(jsonMedicineBackup, json);
        checkBackup(jsonReminderEventBackup, json);
    }

    private <T> void checkBackup(JSONBackup<T> backup, String json) {
        List<T> parsedData = backup.parseBackup(json);
        MedicineRepository medicineRepository = mock(MedicineRepository.class);
        if (parsedData != null) {
            backup.applyBackup(parsedData, medicineRepository);
        }
    }
}

File: app/src/test/java/com/futsch1/medtimer/JSONBackupFuzzTest.kt
package com.futsch1.medtimer

import com.futsch1.medtimer.core.domain.repository.BackupRepository
import com.futsch1.medtimer.database.backup.JSONBackup
import com.futsch1.medtimer.database.backup.JSONMedicineBackup
import com.futsch1.medtimer.database.backup.JSONReminderEventBackup
import kotlinx.coroutines.runBlocking
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.mockito.Mockito

@RunWith(Parameterized::class)
class JSONBackupFuzzTest(private val json: String) {

    companion object {
        @JvmStatic
        @Parameterized.Parameters(name = "{0}")
        fun inputs() = listOf(
            "",
            "{}",
            "[]",
            "null",
            """{"invalid":}""",
            "normal string",
            """{"medicines":[]}"""
        )
    }

    @Test
    fun fuzzTestMedicineBackup() {
        val backupRepository = Mockito.mock<BackupRepository>()

        checkBackup(JSONMedicineBackup(backupRepository), json)
        checkBackup(JSONReminderEventBackup(backupRepository), json)
    }

    private fun <T> checkBackup(backup: JSONBackup<T>, json: String) {
        val parsedData = backup.parseBackup(json)
        if (parsedData != null) {
            runBlocking { backup.applyBackup(parsedData) }
        }
    }
}
